File: Python/AppNotes/AN24_06.py
# Range-Doppler processing for a single channel


# (1) Connect to Radarbook2 and ADF24 Frontend
# (2) Enable Supply
# (3) Configure RX
# (4) Configure TX
# (5) Start Measurements
# (6) Configure signal processing
# (7) Calculate range-Doppler processing for single channel

import  Class.TinyRad as TinyRad
import  time as time
import  numpy as np
from    pyqtgraph.Qt import QtGui, QtCore
import  pyqtgraph as pg


# Configure script
Disp_FrmNr = 1;
Disp_TimSig = 1;      # display time signals
Disp_RP = 0;      # display range profile
Disp_RD = 1;      # display range-Doppler map


c0 = 3e8;


if Disp_TimSig > 0 or Disp_RP > 0 or Disp_RD > 0:
    App = QtGui.QApplication([])

if Disp_TimSig > 0:
    

    WinTim = pg.GraphicsWindow(title="Time signals")
    WinTim.setBackground((255, 255, 255))
    WinTim.resize(1000,600)

    PltTim = WinTim.addPlot(title="TimSig", col=0, row=0)
    PltTim.showGrid(x=True, y=True)

if Disp_RP > 0:
    WinRP = pg.GraphicsWindow(title="Range Profile")
    WinRP.setBackground((255, 255, 255))
    WinRP.resize(1000,600)

    PltRP = WinRP.addPlot(title="Range", col=0, row=0)
    PltRP.showGrid(x=True, y=True)

if Disp_RD:
    View = pg.PlotItem(title='Cross Range Plot')
    View.setLabel('left', 'R', units='m')
    View.setLabel('bottom', 'v', units='m/s')

    Img = pg.ImageView(view=View)
    Img.show()
    Img.ui.roiBtn.hide()
    Img.ui.menuBtn.hide()
    #Img.ui.histogram.hide()
    Img.getHistogramWidget().gradient.loadPreset('flame')

#--------------------------------------------------------------------------
# Setup Connection
#--------------------------------------------------------------------------
Brd = TinyRad.TinyRad()

Brd.BrdRst()

#--------------------------------------------------------------------------
# Software Version
#--------------------------------------------------------------------------
Brd.BrdDispSwVers()

#--------------------------------------------------------------------------
# Configure Receiver
#--------------------------------------------------------------------------
Brd.RfRxEna()
TxPwr =   100

#--------------------------------------------------------------------------
# Configure Transmitter (Antenna 0 - 2, Pwr 0 - 100)
#--------------------------------------------------------------------------
Brd.RfTxEna(2, TxPwr)

#--------------------------------------------------------------------------
# Configure Measurements
# Cfg.Perd: time between measuremnets: must be greater than 1 us*N + 10us
# Cfg.N: number of samples collected: 1e66 * TRampUp = N; if N is smaller
# only the first part of the ramp is sampled; if N is larger than the 
# 
#--------------------------------------------------------------------------
dCfg = dict()
dCfg['fStrt'] = 24.00e9
dCfg['fStop'] = 24.25e9
dCfg['TRampUp'] = 128e-6 
dCfg['Perd'] = 152e-6
dCfg['N'] = 128
dCfg['Seq'] = [1]
dCfg['CycSiz'] = 2
dCfg['FrmSiz'] = 100
dCfg['FrmMeasSiz'] = 32

Brd.RfMeas(dCfg)

#--------------------------------------------------------------------------
# Read actual configuration
#--------------------------------------------------------------------------
N = int(Brd.Get('N'))
NrChn = int(Brd.Get('NrChn'))
fs = Brd.Get('fs')
N = int(Brd.Get('N'))
Np = int(dCfg['FrmMeasSiz'])
#--------------------------------------------------------------------------
# Check TCP/IP data rate:
# 16 Bit * Number of Enabled Channels * Number of Samples are measureed in
# the interval TInt. If the data rate is too high, than frames can be losed
#--------------------------------------------------------------------------
DataRate        =   16*NrChn*N*dCfg['FrmMeasSiz']/(dCfg['FrmSiz']*dCfg['Perd'])
print('DataRate: ', (DataRate/1e6), ' MBit/s')

#--------------------------------------------------------------------------
# Configure Signal processing
#--------------------------------------------------------------------------
# Processing of range profile
Win2D = Brd.hanning(N-1,int(Np))
ScaWin = sum(Win2D[:,0])
NFFT = 2**10
NFFTVel = 2**9
kf = (dCfg['fStop'] - dCfg['fStrt'])/dCfg['TRampUp']
vRange = np.arange(NFFT//2)/NFFT*fs*c0/(2*kf)
fc = (dCfg['fStop'] + dCfg['fStrt'])/2

RMin = 1
RMax = 100
RMinIdx = np.argmin(np.abs(vRange - RMin))
RMaxIdx = np.argmin(np.abs(vRange - RMax))
vRangeExt = vRange[RMinIdx:RMaxIdx]

WinVel2D = Brd.hanning(int(Np), len(vRangeExt))
ScaWinVel = sum(WinVel2D[:,0])
WinVel2D = WinVel2D.transpose()

vFreqVel = np.arange(-NFFTVel//2,NFFTVel//2)/NFFTVel*(1/dCfg['Perd'])
vVel = vFreqVel*c0/(2*fc)

#--------------------------------------------------------------------------
# Generate Curves for plots
#--------------------------------------------------------------------------
if Disp_TimSig:
    n = np.arange(int(N))
    CurveTim = []
    for IdxFrm in np.arange(Np):
        CurveTim.append(PltTim.plot(pen=pg.intColor(IdxFrm, hues=Np)))

if Disp_RP:
    CurveRP = []
    for IdxFrm in np.arange(Np):
        CurveRP.append(PltRP.plot(pen=pg.intColor(IdxFrm, hues=Np)))   


#--------------------------------------------------------------------------
# Measure and calculate Range Doppler Map
#--------------------------------------------------------------------------
# Select channel to be processed. ChnSel <= NrChn
ChnSel = 0

for Cycles in range(0, 1000):
    
    Data = Brd.BrdGetData()
    if Disp_FrmNr  > 0:
        FrmCntr     =   Data[0,:]
        print("FrmCntr: ", FrmCntr)

    # Reshape measurement data for range doppler processing
    # 2D array with [N, NLoop]
    MeasChn = np.reshape(Data[:,ChnSel],(N, Np), order='F')

    # Calculate range profiles and display them
    RP = 2*np.fft.fft(MeasChn[1:,:]*Win2D, n=NFFT, axis=0)/ScaWin*Brd.FuSca
    RP = RP[RMinIdx:RMaxIdx,:]
    
      
    if Disp_TimSig > 0:
        # Display time domain signals
        for IdxFrm in np.arange(Np):
            CurveTim[IdxFrm].setData(n[1:],MeasChn[1:,IdxFrm])
    
    if Disp_RP > 0:
        for IdxFrm in np.arange(Np):
            CurveRP[IdxFrm].setData(vRangeExt,20*np.log10(np.abs(RP[:,IdxFrm])))
    
    if Disp_RD > 0:
        # Display range doppler map
        RD = np.fft.fftshift(np.fft.fft(RP*WinVel2D, NFFTVel, axis=1)/ScaWinVel, axes=1)
        RDMap = 20*np.log10(np.abs(RD))
        Img.setImage(RDMap.transpose(), pos=[vVel[0],RMin], scale=[(vVel[-1] - vVel[0])/NFFTVel,(RMax - RMin)/len(vRangeExt)]) 
        
        View.setAspectLocked(False)     
    pg.QtGui.QApplication.processEvents()




del Brd

File: MatlabRadServe/TinyRad/AppNotes/AN24_06.m
% AN24_06 -- Range-Doppler basics
% Evaluate the range-Doppler map for a single channel
clear all;
close all;

% (1) Connect to DemoRad: Check if Brd exists: Problem with USB driver
% (3) Configure RX
% (4) Configure TX
% (5) Start Measurements
% (6) Configure calculation of range profile and range doppler map for
% channel 1

% Configure script
Disp_FrmNr          =   1;
Disp_TimSig         =   1;      % display time signals
Disp_RP             =   0;      % display range profile
Disp_RD             =   1;      % display range-Doppler map
c0 = 299792458;
%--------------------------------------------------------------------------
% Include all necessary directories
%--------------------------------------------------------------------------
CurPath = pwd();
addpath([CurPath,'/../../UsbMex']);
addpath([CurPath,'/../../Class']);

%--------------------------------------------------------------------------
% Configure Script
%--------------------------------------------------------------------------
% ConSel = 0; Use USB Mex file to communicate directly with TinyRad
% ConSel > 0; Use RadServe to communicate with TinyRad; 
ConSel = 1;

%--------------------------------------------------------------------------
% Setup Connection: open board object
%--------------------------------------------------------------------------
if ConSel > 0
    Brd     =   TinyRad('RadServe', '127.0.0.1');
else
    Brd     =   TinyRad();
end

Brd.BrdRst();

%--------------------------------------------------------------------------
% Software Version
%--------------------------------------------------------------------------
Brd.BrdDispSwVers();

%--------------------------------------------------------------------------
% Configure Receiver
%--------------------------------------------------------------------------
Brd.RfRxEna();
TxPwr           =   80;

%--------------------------------------------------------------------------
% Configure Transmitter (Antenna 0 - 2, Pwr 0 - 100)
%--------------------------------------------------------------------------
Brd.RfTxEna(1, TxPwr);

%--------------------------------------------------------------------------
% Configure Up-Chirp and timing for the measurements
%--------------------------------------------------------------------------
Cfg.fStrt       =   24.00e9;                    %   Start frequency   
Cfg.fStop       =   24.25e9;                    %   Stop frequency
Cfg.TRampUp     =   512e-6;                     %   UpChirp duration
Cfg.Perd        =   0.6e-3;                     %   Period between measurements
Cfg.N           =   512;                        %   Number of samples taken at start of chirp 
Cfg.Seq         =   [1];                        %   Antenna transmit sequence
Cfg.CycSiz      =   2;                          %   Number of buffers in the acquisition framework 2
Cfg.FrmSiz      =   400;                       %   Number of chirp sequences for one measurement cycle
Cfg.FrmMeasSiz  =   64;                         %   Number of chirps sequences for collecting IF data


Brd.RfMeas(Cfg);

%--------------------------------------------------------------------------
% Read actual configuration
%--------------------------------------------------------------------------
NrChn           =   Brd.Get('NrChn');
N               =   Brd.Get('N');
fs              =   Brd.Get('fs');

%--------------------------------------------------------------------------
% Configure Signal Processing
%--------------------------------------------------------------------------
% Processing of range profile
Win2D           =   Brd.hanning(Cfg.N-1,Cfg.FrmMeasSiz);
ScaWin          =   sum(Win2D(:,1));
NFFT            =   2^12;
NFFTVel         =   2^12;
kf              =   (Cfg.fStop - Cfg.fStrt)/Cfg.TRampUp;
vRange          =   [0:NFFT-1].'./NFFT.*fs.*c0/(2.*kf);
fc              =   (Cfg.fStop + Cfg.fStrt)/2;

RMin            =   0;
RMax            =   100;

[Val RMinIdx]   =   min(abs(vRange - RMin));
[Val RMaxIdx]   =   min(abs(vRange - RMax));
vRangeExt       =   vRange(RMinIdx:RMaxIdx);

WinVel          =   Brd.hanning(Cfg.FrmMeasSiz);
ScaWinVel       =   sum(WinVel);
WinVel2D        =   repmat(WinVel.',numel(vRangeExt),1);

vFreqVel        =   [-NFFTVel./2:NFFTVel./2-1].'./NFFTVel.*(1/Cfg.Perd);
vVel            =   vFreqVel*c0/(2.*fc); 

%--------------------------------------------------------------------------
% Measure and calculate range-Doppler Map
%--------------------------------------------------------------------------
% Select channel to be processed. ChnSel <= NrChn
ChnSel          =   1;

for MeasIdx     =   1:100
    
    Data        =   Brd.BrdGetData();
    if Disp_FrmNr  > 0
        disp(num2str(Data(1,:)))
    end
    % Reshape measurement data for range doppler processing
    % 2D array with [N, NLoop]
    MeasChn   =     reshape(Data(:,ChnSel),Cfg.N, Cfg.FrmMeasSiz);
    
    % Calculate range profile including calibration
    RP          =   fft(MeasChn(2:end,:).*Win2D,NFFT,1).*Brd.FuSca/ScaWin;
    RPExt       =   RP(RMinIdx:RMaxIdx,:);    

    RD          =   fft(RPExt.*WinVel2D, NFFTVel, 2)./ScaWinVel;
    RD          =   fftshift(RD, 2);
    
    if Disp_TimSig > 0
        % Display time signals
        figure(1)
        plot(MeasChn(2:end,:));
        grid on;
        xlabel('n ( )');
        ylabel('u (LSB)');     
    end
    
    if Disp_RP > 0
        % Display range profile
        figure(2)
        plot(vRangeExt, 20.*log10(abs(RPExt)));
        grid on;
        xlabel('R (m)');
        ylabel('X (dBV)');
        axis([vRangeExt(1) vRangeExt(end) -120 -40])
    end
    
    if Disp_RD > 0
        % Display range doppler map
        figure(3)
        imagesc(vVel, vRangeExt, 20.*log10(abs(RD)));
        grid on;
        xlabel('v (m/s)');
        ylabel('R (m)');
        colormap('jet')
    end

    drawnow()

end

clear Brd